src/models/geograph.py

```diff
@@ -443,27 +450,143 @@ def merge_nodes(
             and self.graph.has_node(final_index)
         ):
             raise ValueError(
-                "`final_index` must not be an existing node not in `node_list`."
+                "`final_index` cannot be an existing node that is not in `node_list`."
             )
         if final_index is None:
-            final_index = self.graph.number_of_nodes
+            final_index = self.df.last_valid_index()
 
-        adjacency_list = set()
+        # Build set of all neighbours of nodes in node_list, excluding the
+        # nodes in node_list.
+        adjacency_set = set()
         for node in node_list:
-            adjacency_list.update(self.graph.neighbors(node))
-        polygon = unary_union(
-            [self.graph.nodes[node]["geometry"] for node in node_list]
-        )
-
+            adjacency_set.update(list(self.graph.neighbors(node)))
+        adjacency_set -= adjacency_set.intersection(node_list)
+        # Build union polygon.
+        polygon = self.df.geometry.iloc[node_list].unary_union
+        # Remove nodes from graph and rows from df
         self.graph.remove_nodes_from(node_list)
+        self.df = self.df.drop(node_list)
```

Comment:
Will this lead to potentially missing indices?
(I'm thinking e.g. `1, 2, 3, 5, 6, 8, ...` ? If yes, this could be a problme with how we check `final_index` above)